src/console.rs

```diff
@@ -0,0 +1,61 @@
+#![allow(dead_code)]
+use std::error::Error;
+
+use crate::{
+	error::{self, ErrorCode},
+	write_to_ipc, IPC_BUFFER_SIZE,
+};
+
+extern "C" {
+	fn console_log(is_string: bool, ipc_offset: usize);
+}
+
+/// Log a string to the console
+pub fn log_str<S: AsRef<str>>(msg: S) {
+	let data = msg.as_ref().as_bytes();
+
+	unsafe {
+		if data.len() >= IPC_BUFFER_SIZE {
+			let error_msg: &[u8] = b"The length of data to be logged to the console exceeds the size of the IPC_BUFFER";
+			write_to_ipc(error_msg);
+			error::exit(ErrorCode::LogDataTooLong, error_msg.len())
+		};
+
+		write_to_ipc(data);
+		console_log(true, data.len());
+	}
+}
+
+/// Log a string to the console
+pub fn log_buf<S: AsRef<[u8]>>(data: S) {
+	let data = data.as_ref();
+
+	unsafe {
```

Comment:
`.wasm` is single-threaded so there can never be any two threads reading or writing from static mutable memory. In Rust, any `static mut` variable is unsafe because of data races, but what are you racing against in a single-threaded application

Reply:
Plus all writing to the IPC is handled via the `write_to_ipc` function, which is safe to execute. The unsafe part of the logging functions is calling the `extern "C" **` functions